### ISSUE TITLE
`cplusplus.Move` treats the left-hand side of an explicit-object move assignment as moved-from

### ISSUE THREAD

**Issue by SekaiArendelle** [llvm/llvm-project#218227]
## Description

The Clang Static Analyzer's `cplusplus.Move` checker appears to model a C++23 explicit-object move-assignment operator incorrectly.

For an operator declared as:

```cpp
auto operator=(this Owner& self, Owner&& other) noexcept -> Owner&;
```

the analyzer marks the explicit object argument (`self`, the left-hand side of the assignment) as moved-from after an expression such as:

```cpp
target = std::move(source);
```

It should mark `source` as moved-from. `target` has just received a new value and remains valid.

This produces false positives when `target` is subsequently moved. It also makes `scan-build --status-bugs` fail otherwise successful builds.

## Minimal reproducer

[godbolt link](https://godbolt.org/z/q9GG4jbqT)

```cpp
#include <utility>

struct Owner {
    int* pointer{};

    constexpr Owner() noexcept = default;

    constexpr Owner(Owner&& other) noexcept
        : pointer{std::exchange(other.pointer, nullptr)} {
    }

    constexpr auto operator=(this Owner& self, Owner&& other) noexcept -> Owner& {
        self.pointer = std::exchange(other.pointer, nullptr);
        return self;
    }
};

struct Holder {
    Owner value;

    constexpr explicit Holder(Owner&& source) noexcept
        : value{std::move(source)} {
    }
};

void reproducer() {
    Owner target{};
    Owner source{};
    target = std::move(source);
    Holder holder{std::move(target)};
}
```

## Command

```sh
clang++ --analyze \
  -Xanalyzer -analyzer-checker=cplusplus.Move \
  -Xanalyzer -analyzer-output=text \
  -std=c++23 repro.cc
```

The same issue is visible when running a CMake build through `scan-build --status-bugs`.

## Actual result

The analyzer reports that `target` is moved-from after the move-assignment expression and warns when it is subsequently moved into `Holder`:

```text
warning: Moved-from object 'target' is moved [cplusplus.Move]
note: Object 'target' is moved
    target = std::move(source);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~
note: Calling constructor for 'Holder'
    Holder holder{std::move(target)};
           ^~~~~~~~~~~~~~~~~~~~~~~~~
note: Moved-from object 'target' is moved
    : value{std::move(source)} {
      ^~~~~~~~~~~~~~~~~~~~~~~~
```

## Expected result

No diagnostic should be emitted for the later move from `target`.

After `target = std::move(source)`, `source` is the moved-from object. The move-assignment operator has assigned a new valid state to `target`.

## Comparison with an implicit-object move-assignment operator

Replacing only the move-assignment operator with the traditional implicit-object form removes the diagnostic:

```cpp
constexpr auto operator=(Owner&& other) noexcept -> Owner& {
    pointer = std::exchange(other.pointer, nullptr);
    return *this;
}
```

This suggests that the problem is specifically in the analyzer's modeling of the explicit object parameter rather than in its general modeling of move assignment.

## Versions observed

- Clang 21.1.8, installed from Ubuntu 26.04 packages and invoked through `scan-build`, reports the false positive in the original project.
- Clang 23.0.0git (`4c4c1db7c69a6fda6cfa6bc6066bb09a433edc89`), invoked directly with `clang++ --analyze` on Windows, reproduces it with the minimal example above.

The implicit-object variant produces no warning under the same Clang 23 analyzer invocation.

**Comment by haoNoQ** [comment]
This might be yet another cornercase where `this` is suddenly added to the list of parameters/arguments so now every tool needs to explicitly filter it out.

And given that the motivation for C++23 explicit object parameters was to make non-static member functions act more like static member functions (with respect to being able to make template substitutions inside the object parameter type), I guess it makes sense that it's the same problem.

(I haven't actually looked into it to confirm.)

**Comment by SekaiArendelle** [comment]
Based on the fact that the explicit obj call wasn't modeled as a CXXInstanceCall, I found another similar issues.

`ContainerModeling::checkPostCall` handles begin/end, modifiers and assignment only under `dyn_cast&lt;CXXInstanceCall&gt;`, so explicit-object container methods are never modeled.
Downstream, `MismatchedIteratorChecker` and invalidation diagnostics that depend on container state also go silent for such types.

[godbolt link](https://godbolt.org/z/7Pv76vW63)

```cpp
struct iterator {};

struct ImplicitContainer {
  iterator begin();
  iterator end();
};

struct ExplicitContainer {
  iterator begin(this ExplicitContainer &);
  iterator end(this ExplicitContainer &);
};

template <typename Container>
long clang_analyzer_container_begin(const Container&);
template <typename Container>
long clang_analyzer_container_end(const Container&);
void clang_analyzer_denote(long, const char*);
void clang_analyzer_express(long);

void implicit_case(ImplicitContainer v0) {
  v0.begin();
  v0.end();

  clang_analyzer_denote(clang_analyzer_container_begin(v0), "$b0");
  clang_analyzer_denote(clang_analyzer_container_end(v0), "$e0");

  clang_analyzer_express(clang_analyzer_container_begin(v0)); // expect: $b0
  clang_analyzer_express(clang_analyzer_container_end(v0));   // expect: $e0
}

void explicit_case(ExplicitContainer v0) {
  v0.begin();
  v0.end();

  clang_analyzer_denote(clang_analyzer_container_begin(v0), "$b1");
  clang_analyzer_denote(clang_analyzer_container_end(v0), "$e1");

  clang_analyzer_express(clang_analyzer_container_begin(v0)); // modeled => $b1, missed => 0
  clang_analyzer_express(clang_analyzer_container_end(v0));   // modeled => $e1, missed => 0
}
```

output:
```
<source>:29:3: warning: $b0 [debug.ExprInspection]
   29 |   clang_analyzer_express(clang_analyzer_container_begin(v0)); // expect: $b0
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:29:3: note: $b0
   29 |   clang_analyzer_express(clang_analyzer_container_begin(v0)); // expect: $b0
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:30:3: warning: $e0 [debug.ExprInspection]
   30 |   clang_analyzer_express(clang_analyzer_container_end(v0));   // expect: $e0
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:30:3: note: $e0
   30 |   clang_analyzer_express(clang_analyzer_container_end(v0));   // expect: $e0
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:37:3: warning: Not a symbol [debug.ExprInspection]
   37 |   clang_analyzer_denote(clang_analyzer_container_begin(v0), "$b1");
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:37:3: note: Not a symbol
   37 |   clang_analyzer_denote(clang_analyzer_container_begin(v0), "$b1");
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:38:3: warning: Not a symbol [debug.ExprInspection]
   38 |   clang_analyzer_denote(clang_analyzer_container_end(v0), "$e1");
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:38:3: note: Not a symbol
   38 |   clang_analyzer_denote(clang_analyzer_container_end(v0), "$e1");
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:40:3: warning: Not a symbol [debug.ExprInspection]
   40 |   clang_analyzer_express(clang_analyzer_container_begin(v0)); // modeled => $b1, missed => 0
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:40:3: note: Not a symbol
   40 |   clang_analyzer_express(clang_analyzer_container_begin(v0)); // modeled => $b1, missed => 0
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:41:3: warning: Not a symbol [debug.ExprInspection]
   41 |   clang_analyzer_express(clang_analyzer_container_end(v0));   // modeled => $e1, missed => 0
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:41:3: note: Not a symbol
   41 |   clang_analyzer_express(clang_analyzer_container_end(v0));   // modeled => $e1, missed => 0
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
6 warnings generated.
```
